include/ktopt_interface/ktopt_planning_context.hpp

```diff
@@ -12,63 +12,87 @@
+using drake::multibody::AddMultibodyPlantSceneGraph;
+using drake::multibody::MultibodyPlant;
+using drake::multibody::PackageMap;
 using drake::multibody::Parser;
 using drake::planning::trajectory_optimization::KinematicTrajectoryOptimization;
 using drake::solvers::Solve;
-using drake::systems::Diagram;
 using drake::systems::Context;
-using drake::multibody::PackageMap;
-using Eigen::VectorXd;
+using drake::systems::Diagram;
+using drake::systems::DiagramBuilder;
+using drake::visualization::ApplyVisualizationConfig;
+using drake::visualization::VisualizationConfig;
+using drake::math::RigidTransformd;
 using Eigen::MatrixXd;
+using Eigen::VectorXd;
+using Eigen::Vector3d;
 using Joints = std::vector<const moveit::core::JointModel*>;
 
 class KTOptPlanningContext : public planning_interface::PlanningContext
 {
 public:
-    KTOptPlanningContext(
-        const std::string& name,
-        const std::string& group_name,
-        const ktopt_interface::Params& params);
-
-    void solve(
-        planning_interface::MotionPlanResponse& res) override;
-    void solve(
-        planning_interface::MotionPlanDetailedResponse& res) override;
+  KTOptPlanningContext(const std::string& name, const std::string& group_name, const ktopt_interface::Params& params);
 
-    bool terminate() override;
-    void clear() override;
+  void solve(planning_interface::MotionPlanResponse& res) override;
+  void solve(planning_interface::MotionPlanDetailedResponse& res) override;
 
-    void setRobotDescription(std::string robot_description);
-    VectorXd toDrakePositions(
-        const moveit::core::RobotState& state,
-        const Joints& joints);
-    void setJointPositions(
-        const VectorXd& values,
-        const Joints& joints,
-        moveit::core::RobotState& state);
-    void setJointVelocities(
-        const VectorXd& values,
-        const Joints& joints,
-        moveit::core::RobotState& state);
+  bool terminate() override;
+  void clear() override;
 
+  void setRobotDescription(std::string robot_description);
+  VectorXd toDrakePositions(const moveit::core::RobotState& state, const Joints& joints);
+  void setJointPositions(const VectorXd& values, const Joints& joints, moveit::core::RobotState& state);
+  void setJointVelocities(const VectorXd& values, const Joints& joints, moveit::core::RobotState& state);
+  void transcribePlanningScene(const planning_scene::PlanningScene& planning_scene,
+                           MultibodyPlant<double>& plant, SceneGraph<double>& scene_graph);
 
 private:
-    const ktopt_interface::Params params_;
-    std::string robot_description_;
+  const ktopt_interface::Params params_;
+  std::string robot_description_;
+
+  // drake related variables
+  SceneGraph<double>* scene_graph_{};
+  MultibodyPlant<double>* plant_{};
+  std::unique_ptr<Diagram<double>> diagram_;
+  std::unique_ptr<Context<double>> diagram_context_;
+  Context<double>* plant_context_{};
+  Context<double>* visualizer_context_{};
```

Comment:
my first suspicion as to why you're getting segfaults is that storing these raw pointers as member variables is actually very memory unsafe and likely something went out of scope.

I think you should actually try removing `scene_graph_`, `plant_`, `plant_context_`, and `visualizer_context_` from getting member variables -- we recently ran into this at work.

The correct usage is to leverage the `diagram_` smart pointer you have to do extractions like these in the function scopes that require them:

```c++
  const auto& plant =
      dynamic_cast<const drake::multibody::MultibodyPlant<double>&>(diagram_->GetSubsystemByName("plant"));
  auto& plant_context = diagram_->GetMutableSubsystemContext(plant, diagram_context_.get());
```

I know it seems like a lot, but I think this is how Drake is intended to be used based on our tireless comb through the source code in lieu of there being very little examples to borrow from.

Reply:
Thanks for this, I did see them use this in their examples initially, but I went for whatever was compiling. I'll refactor accordingly and test.